### DISCUSSION TITLE
Lazy question! Is anyome here using simple "video capture" from a webcam?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#206] [Q&A] [answered]
It was great fun and huge impact to use in class the Processing `video` library in class, and then with very few clicks on the Processing IDE, the `OpenCV for Processing`  by [Greg Borenstein](http://gregborenstein.com/) & Florian Bruggisser.

Sadly, video & video capture in Processing Python mode started to fail in different platforms (much like Sound and Serial*), and, well, I don't want to use Processing Python mode in class anymore.
> \*  I was able to re-write my Serial/Firmata projects with pyFirmata and it was ok.
> I know there are people working on *Sound*.

So, before I put energy on this, has anyone tried something on this direction?

- reference: [ASCII camera example with OpenCV capture + PySympleGUI](https://gist.github.com/villares/76313507fca73ffed72ec08220e78278)

**Comment by villares** [comment, marked as answer]
![image](https://user-images.githubusercontent.com/3694604/206069716-44c4fa7b-3f61-4a7a-a3cf-46c45ef78614.png)

This seems to work...

```python
import cv2
import numpy as np
#from PIL import Image
from py5 import create_image_from_numpy

cap = cv2.VideoCapture(0)

def setup():
    size(640, 480)

def draw():
    ret, frame = cap.read()
    #pil_img = Image.fromarray(frame)     # from npa to PIL.Image
    ## if you want to use PIL manipulation here
    #img_as_np_array = np.array(pil_img)  # back to npa
    img_as_np_array = frame
    gray = cv2.cvtColor(img_as_np_array, cv2.COLOR_BGR2GRAY)
    edges = cv2.Canny(gray, 100, 80)
    # blended_gray = cv2.addWeighted(gray, 0.6, edges, 0.4, 0)
    edges_rgb_npa = cv2.cvtColor(edges, cv2.COLOR_GRAY2BGR)
    blended_rgb_npa = cv2.addWeighted(img_as_np_array, 0.6,
                                      edges_rgb_npa, 0.4, 0)
    py5_img = create_image_from_numpy(blended_rgb_npa, 'RGB')
    image(py5_img, 0, 0)
     
def exiting():
    print('over and out')
    cap.release()
```

**Reply by hx2A** [reply]
This is a clear and simple solution!

What is the frame rate of this Sketch? I've used opencv to do webcam stuff with py5 and found it to work very well. One thing you might notice is that the frame rate of the Sketch matches the frame rate of the webcam. This is because the `cap.read()` call will block until it reads the next frame from the video source. This forces your Sketch to be in sync with the webcam. If you don't want this, you can use threading instead, reading the webcam from a repeating thread.

One way you can improve this is by using the `dst` param for `create_image_from_numpy()`:

http://py5coding.org/reference/sketch_create_image_from_numpy.html

It is better to recycle the Py5Image object and not create a new one for every frame. The code would look like this:

```python

py5_img = None

def draw():
    global py5_img
    ...
    py5_img = create_image_from_numpy(blended_rgb_npa, 'RGB', dst=py5_img)
```

This will use the same Py5Image object for every frame.

**Reply by villares** [reply]
Works great, but I don't see that much difference (both ways about 30 FPS), but then maybe my computer & camera are slow...

```python
# Updated using hx2A's tip from:
# https://github.com/py5coding/py5generator/discussions/206#discussioncomment-4329841

import cv2
import numpy as np
from py5 import create_image_from_numpy

cap = cv2.VideoCapture(0)
py5_img = None

def setup():
    size(640, 480)

def draw():
    global py5_img
    if frame_count % 30 == 0:
        window_title(f'FR: {get_frame_rate()}')
    ret, frame = cap.read()  # frame is a numpy array
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    edges = cv2.Canny(gray, 100, 80)
    #blended_gray = cv2.addWeighted(gray, 0.6, edges, 0.4, 0)
    edges_bgr_npa = cv2.cvtColor(edges, cv2.COLOR_GRAY2BGR)
    blended_bgr_npa = cv2.addWeighted(frame, 0.6,
                                      edges_bgr_npa , 0.4, 0)
    blended_rgb_npa = cv2.cvtColor(blended_bgr_npa , cv2.COLOR_BGR2RGBA)
    py5_img = create_image_from_numpy(blended_rgb_npa, 'RGB', dst=py5_img)
    # display image
    image(py5_img, 0, 0)
     
def exiting():
    print('over and out')
    cap.release()
```

Updated with: ` cv2.cvtColor(blended_bgr_npa , cv2.COLOR_BGR2RGBA)`

**Reply by hx2A** [reply]
> Works great, but I don't see that much difference (both ways about 30 FPS), but then maybe my computer & camera are slow...

Your webcam must be 30 fps. If you were using threading and running the Sketch at 60 fps, you might see a small difference.

In any case, that performance feature is there for when you need it.

**Reply by villares** [reply]
Cheers @hx2A. I now learned that OpenCV captured frames (cv2) are always BGR. So I had to add a BGR to RGB conversion. Something that might be a py5 issue: when I tried using "BGR" as an argument in `create_image_from_numpy()` it failed silently returning a blank image. Could you please add a warning? (Or maybe add these alternative modes...?)

**Reply by hx2A** [reply]
> Something that might be a py5 issue: when I tried using "BGR" as an argument in `create_image_from_numpy()` it failed silently returning a blank image. Could you please add a warning? (Or maybe add these alternative modes...?)

This issue came up a month ago in a conversation with @marziabil and I recall I added support for `BGR` ordering. What version of py5 are you using? Are you using the test version of py5 I gave you or the latest release of py5?

**Reply by villares** [reply]
It was at my Sesc job's computer, so it must have been an older version. I'll test on the dev/test version then! Very cool!
